zcash_client_sqlite/src/wallet/sapling.rs

```diff
@@ -132,14 +132,33 @@ pub(crate) fn get_spendable_sapling_notes(
     anchor_height: BlockHeight,
     exclude: &[ReceivedNoteId],
 ) -> Result<Vec<ReceivedSaplingNote<ReceivedNoteId>>, SqliteClientError> {
+    let mut stmt_unscanned_tip = conn.prepare_cached(
+        "SELECT 1 FROM v_sapling_shard_unscanned_ranges unscanned
+         WHERE unscanned.shard_index IN (SELECT MAX(shard_index) FROM sapling_tree_shards)
+         AND unscanned.subtree_start_height <= :anchor_height",
+    )?;
+    let mut unscanned =
+        stmt_unscanned_tip.query(named_params![":anchor_height": &u32::from(anchor_height),])?;
+    if unscanned.next()?.is_some() {
+        // if the tip shard has unscanned ranges below the anchor height, none of our notes can be
+        // spent
+        return Ok(vec![]);
+    }
+
     let mut stmt_select_notes = conn.prepare_cached(
         "SELECT id_note, diversifier, value, rcm, commitment_tree_position
          FROM sapling_received_notes
          INNER JOIN transactions ON transactions.id_tx = sapling_received_notes.tx
          WHERE account = :account
          AND spent IS NULL
          AND transactions.block <= :anchor_height
-         AND id_note NOT IN rarray(:exclude)",
+         AND id_note NOT IN rarray(:exclude)
+         AND NOT EXISTS (
+            SELECT 1 FROM v_sapling_shard_unscanned_ranges unscanned
+            WHERE sapling_received_notes.commitment_tree_position >= unscanned.start_position
+            AND sapling_received_notes.commitment_tree_position < unscanned.end_position_exclusive
```

Comment:
Consider `WHERE sapling_received_notes.commitment_tree_position BETWEEN unscanned.start_position AND unscanned.end_position_exclusive - 1`.

(`BETWEEN` does the correct thing, i.e. return false, when its end value is less than its start value.)